pkg/spdk/replica.go

```diff
@@ -372,39 +380,55 @@ func (r *Replica) Create(spdkClient *spdkclient.Client, port int32, exposeRequir
 	if err != nil {
 		return nil, err
 	}
+	r.IP = podIP
+
+	r.PortStart, r.PortEnd, err = superiorPortAllocator.AllocateRange(types.DefaultReplicaReservedPortCount)
+	if err != nil {
+		return nil, err
+	}
+	// Always reserved the 1st port for replica expose and the rest for rebuilding
+	r.portAllocator = util.NewBitmap(r.PortStart+1, r.PortEnd)
+
 	if exposeRequired {
-		if err := spdkClient.StartExposeBdev(helpertypes.GetNQN(r.Name), r.UUID, podIP, string(r.Port)); err != nil {
+		if err := spdkClient.StartExposeBdev(helpertypes.GetNQN(r.Name), r.UUID, podIP, string(r.PortStart)); err != nil {
 			return nil, err
 		}
+		r.IsExposed = true
 	}
-	r.IP = podIP
-	r.Port = port
 	r.State = ReplicaStateStarted
 
 	return ServiceReplicaToProtoReplica(r), nil
 }
 
-func (r *Replica) Delete(spdkClient *spdkclient.Client, cleanupRequired bool) (err error) {
+func (r *Replica) Delete(spdkClient *spdkclient.Client, cleanupRequired bool, superiorPortAllocator *util.Bitmap) (err error) {
 	r.Lock()
 	defer r.Unlock()
 
-	exposeStopped := false
 	defer func() {
 		if err != nil {
 			r.State = ReplicaStateError
 		} else {
 			r.State = ReplicaStateStopped
 		}
-		if exposeStopped {
-			r.IP = ""
-			r.Port = 0
+		// The port can be released once the rebuilding and expose are stopped
+		if !r.IsExposed && r.PortStart != 0 {
+			if releaseErr := superiorPortAllocator.ReleaseRange(r.PortStart, r.PortEnd); releaseErr != nil {
+				r.log.Errorf("Failed to release port %d to %d at the end of replica deletion: %v", r.PortStart, r.PortEnd, releaseErr)
+				return
+			}
```

Comment:
In this case, it will be orphaned replica? how can these ports be reused eventually?

Reply:
Ideally, the monitor will periodically list all lvols and build cache for each RW lvol, which is also the missing part for the disk create/add API.
I will handle it after the demo.